### PR TITLE
Chunked prefill support

### PR DESCRIPTION
**Pull request by hnyls2002**
Thank you for your contribution, we really appreciate it. The following instructions will help improve your pull request and make it easier to receive feedback. If there are any items you don't understand, don't worry. Just submit the pull request and ask the maintainers for help.

## Motivation

Support chunked prefill.

## Modification

1. Add an `--chunked-prefill-size` parameter.
2. Add an inflight request field in tp worker.

## Checklist

1. Ensure pre-commit `pre-commit run --all-files` or other linting tools are used to fix potential lint issues.
3. Confirm that modifications are covered by complete unit tests. If not, please add more unit tests for correctness.
4. Modify documentation as needed, such as docstrings or example tutorials.

### CODE REVIEW

**Comment by hnyls2002** [comment]
Also thanks @vikranth22446 for his help with chunked prefill.

File: python/sglang/srt/server_args.py [hunk 1]
```diff
@@ -315,6 +318,14 @@ def add_cli_args(parser: argparse.ArgumentParser):
         )
         parser.add_argument("--node-rank", type=int, help="The node rank.")
 
+        # Chunked prefill
+        parser.add_argument(
+            "--chunked-prefill-size",
+            type=int,
+            default=0,
+            help="The size of the chunked prefill.",
```
Comment (Ying1123) [inline python/sglang/srt/server_args.py]: make default `None`

File: python/sglang/srt/server_args.py [hunk 2]
```diff
@@ -393,6 +404,14 @@ def check_server_args(self):
             self.dp_size > 1 and self.node_rank is not None
         ), "multi-node data parallel is not supported"
 
+        assert not (
+            self.chunked_prefill_size > 0 and self.disable_radix_cache
+        ), "chunked prefill is not supported with radix cache disabled currently"
+
+    def post_server_args(self):
```
Comment (Ying1123) [inline python/sglang/srt/server_args.py]: why not using the existing `__post_init__`

File: python/sglang/srt/server.py [hunk 3]
```diff
@@ -175,12 +175,46 @@ def _set_torch_compile_config():
     torch._dynamo.config.accumulated_cache_size_limit = 256
 
 
+def set_envs_and_config(server_args: ServerArgs):
```
Comment (Ying1123) [inline python/sglang/srt/server.py:178]: great refactor

File: python/sglang/srt/managers/controller/tp_worker.py [hunk 4]
```diff
@@ -711,8 +768,20 @@ def handle_finished_requests(self, batch: Batch):
             else:
                 batch.reqs = []
 
+    def filter_out_inflight(self, batch: Batch):
+        if self.current_inflight_req is None:
+            return
+
+        unfinished_indices = list(range(len(batch.reqs)))
+        for i, req in enumerate(batch.reqs):
```
Comment (Ying1123) [inline python/sglang/srt/managers/controller/tp_worker.py]: Could you try something like `batch.reqs.index(self.current_inflight_req)` to find the index to remove?

File: python/sglang/srt/managers/controller/tp_worker.py [hunk 5]
```diff
@@ -711,8 +768,20 @@ def handle_finished_requests(self, batch: Batch):
...
+            if req is self.current_inflight_req:
+                unfinished_indices.remove(i)
+                break
+
+        batch.filter_batch(unfinished_indices)
```
Comment (Ying1123) [inline python/sglang/srt/managers/controller/tp_worker.py:784]: Here, we only drop one index. Maybe we can have a specialized version for this.

**Review by Ying1123 — CHANGES_REQUESTED** [review]

File: python/sglang/srt/managers/controller/tp_worker.py (hunk 4 above)
Comment (hnyls2002) [reply python/sglang/srt/managers/controller/tp_worker.py]: sorry, I was a little confused about the judging method between `is` and `==`

File: python/sglang/srt/managers/controller/tp_worker.py [hunk 6]
```diff
@@ -367,7 +372,33 @@ def get_new_prefill_batch(self) -> Optional[Batch]:
                 ]
             )
 
-        for req in self.forward_queue:
+        # Handle the current inflight request
+        take_inflight = 0
```
Comment (Ying1123) [inline python/sglang/srt/managers/controller/tp_worker.py:377]: This function starts to get longer. Consider splitting the body and conditions into multiple subfunctions with good names.

File: python/sglang/srt/managers/controller/tp_worker.py [hunk 7]
```diff
@@ -711,8 +768,17 @@ def handle_finished_requests(self, batch: Batch):
             else:
                 batch.reqs = []
 
+    def filter_out_inflight(self, batch: Batch):
+        if self.current_inflight_req is None:
+            return
+
+        unfinished_indices = list(range(len(batch.reqs)))
+        unfinished_indices.remove(batch.reqs.index(self.current_inflight_req))
```
Comment (Ying1123) [inline python/sglang/srt/managers/controller/tp_worker.py:782]:
```python
to_remove = batch.reqs.index(self.current_inflight_req)
unfinished_indices = [x for x in range(len(batch.reqs)) if x != to_remove]
```

File: python/sglang/srt/server_args.py [hunk 8]
```diff
@@ -65,6 +65,9 @@ class ServerArgs:
     dp_size: int = 1
     load_balance_method: str = "round_robin"
 
+    # Chunked Prefill
+    chunked_prefill_size: int = None
```
Comment (Ying1123) [inline python/sglang/srt/server_args.py]:
```suggestion
    chunked_prefill_size: Optional[int] = None
```

File: python/sglang/srt/server_args.py [hunk 9]
```diff
@@ -315,6 +320,14 @@ def add_cli_args(parser: argparse.ArgumentParser):
         )
         parser.add_argument("--node-rank", type=int, help="The node rank.")
 
+        # Chunked prefill
+        parser.add_argument(
+            "--chunked-prefill-size",
+            type=int,
+            default=None,
```
Comment (Ying1123) [inline python/sglang/srt/server_args.py]:
```suggestion
            default=ServerArgs.chunked_prefill_size,
```

File: python/sglang/srt/managers/controller/tp_worker.py (hunk 5 above)
Comment (hnyls2002) [reply python/sglang/srt/managers/controller/tp_worker.py:784]: I have no idea how to simplify it yet. I will try in the next PR.

File: python/sglang/srt/managers/controller/tp_worker.py (hunk 6 above)
Comment (hnyls2002) [reply python/sglang/srt/managers/controller/tp_worker.py:377]: Sure, there are still some functions to be implemented, I will organize it in the next PR.

**Pull request merged**